Introduce static class hierarchy.

mboes opened the pull request:
`Closure` is not a functor, since we cannot map arbitrary functions
over it. But it sure looks like one, and an applicative one at that.
What we can do is map "static" functions over it (i.e. a `Closure`).
`Closure` is not just an applicative functor, it's also a monad, as
well as a comonad, if again we limit the function space to those
functions that can be statically pointed to. In fact an entire
hierarchy of classes mirroring the standard classes can be defined.

This PR defines such a class hierarchy in the category of static
functions.

cc @edsko

mboes committed:
Introduce static class hierarchy.

`Closure` is not a functor, since we cannot map arbitrary functions
over it. But it sure looks like one, and an applicative one at that.
What we can do is map "static" functions over it (i.e. a `Closure`).
`Closure` is not just an applicative functor, it's also a monad, as
well as a comonad, if again we limit the function space to those
functions that can be statically pointed to. In fact an entire
hierarchy of classes mirroring the standard classes can be defined.

aspiwack commented on src/Control/Distributed/Closure/Class.hs:
```diff
@@ -0,0 +1,81 @@
+class StaticApply f => StaticApplicative f where
+  staticPure :: Typeable a => a -> f a
+
+class StaticApply m => StaticBind m where
```
I'm not sure `Bind` is related to `Apply`. Without a `pure` we seem to be lost.

I'm wondering though, would it make sense to simply take `pure` be `pure :: Closure a -> f a` and not deal with so many type classes (obviously that's realisable when `f = Closure`, do we have examples where we don't have such a `pure`?).

mboes replied:
Well, if we're in the category of static functions between arbitrary types, then we should allow lifting values of arbitrary type. And FWIW `class Apply m => Bind m` in `semigroupoids`.

aspiwack replied:
Ok. I don't think there needs to be a link `Apply => Bind` but let's keep the same relations a `semigroupoids`.

An also, now that I've thought a bit more, what we are actually defining here are endofunctors in the category of static functions. We've just massaged ignored some of the `Closure` for the sake of convenience. If we wrote everything as endofunctor, we would have, I think the following types:
```haskell
map :: Closure (a->b) -> Closure(f a -> f b)
pure :: Closure (a -> f a)
join :: Closure (f (f a) -> a)
bind :: Closure (a -> f b) -> Closure (f a -> f b)
ap :: Closure (f (a -> b) -> f a -> f b)
```

So I'm convinced with the types now. Maybe these are also of interest? Together with a `Category` instance on `Closure (a -> b)`? (we can also implement `(***)` if I'm not mistaken, but not `arr`, but maybe we should add a `StaticArrow` type-class as well?).

mboes replied:
I wrote the laws down, and the only one that breaks is this one:

```haskell
f <*> x = f >>= (<$> x)
```

This law, transposed, says you should be able to implement `staticApply` using `staticBind` and `staticMap` alone. But actually, with the current signature of `staticBind`, I don't think we can implement `staticApply`.

Yet the current type for `staticBind` is pretty clearly what we want. It's the signature we'd want e.g. for `RDD`, where `staticBind` would naturally be `RDD.flatMap`. OTOH `RDD` is a good example of the `staticApply` signature not making much sense.

This raises the question: in the category of static functions, should all monads be applicatives like they are in the category of functions? I guess this ties back to your point about dropping `StaticApply` as a superclass of `StaticBind`...

aspiwack commented on src/Control/Distributed/Closure/Class.hs:
```diff
@@ -0,0 +1,81 @@
+class (StaticApplicative m, StaticBind m) => StaticMonad m
+instance (StaticApplicative m, StaticBind m) => StaticMonad m
+
+staticReturn :: (StaticApplicative m, Typeable a) => a -> m a
```
Meh. Having both `return` and `pure` in Haskell's base is an accident of history. Could we just choose one name and leave the other out of this library?

mboes replied:
The idea here isn't to rewrite Haskell history, except where we have to (in particular the semigroupoid type classes are a necessity for us). I did move `staticReturn` out of being a class method. But principle of least surprise and all, I think it makes sense to retain the symmetry with standard type classes and have both names available.

aspiwack commented on src/Control/Distributed/Closure/Internal.hs:
```diff
@@ -167,4 +167,5 @@ cap closf closx = Ap closf closx
 cmap :: Typeable a => StaticPtr (a -> b) -> Closure a -> Closure b
-cmap sptr = cap (Closure (deRefStaticPtr sptr) (StaticPtr sptr))
+cmap sf = cap (closure sf)
+{-# DEPRECATED cmap "Use staticMap instead." #-}
```
Or `cap` for that matter.

Which, I guess, is both `staticMap` _and_ `staticApply`.

aspiwack reviewed (commented):
This is starting to take quite a nice shape.

We should probably try and write a short RFC to visualise the instances of these type classes that we have in mind, to validate the design.

Cf my first comment for a possible variant that I do find appealing.

aspiwack commented:
Another thing we should endeavour to do is spell out the laws that these things should verify. Just to make sure that the definitions do make sense.

mboes committed:
StaticExtend class.

semigroupoid version of StaticComonad. For consistency with the rest
of the class hierarchy.

mboes committed:
Fix signature of staticApply.

Make it consistent with `staticApply (staticPure f) x = staticMap f x`
law.

mboes committed:
Laws for StaticApply and StaticBind.

mboes committed:
Add capDup combinator. Useful for stating Apply/Bind laws.

mboes committed:
Fix Apply/Bind laws.

mboes committed:
Add StaticBifunctor, StaticProfunctor, StaticStrong and StaticChoice.

Thes are cardinal points in a lattice of profunctor optics, as
explained in "Profunctor Optics: modular data accessors" by Pickering
et al.

mboes committed:
Define type class hierarchy in separate modules. Remove StaticBifunctor.

We split type classes into their own modules that mirror the standard
module hierarchy, with `.Static` appended to the end of the names of
each module.

aspiwack commented on src/Data/Functor/Static.hs:
```diff
@@ -0,0 +1,32 @@
+-- | 'Closure' is not a functor, since we cannot map arbitrary functions over
+-- it. But it sure looks like one, and an applicative one at that. What we can
+-- do is map /static pointers to/ arbitrary functions over it (or in general,
```
Minor: the reference to static pointers here seems outdated.

mboes replied:
I mention `Closure` in the same sentence. Anything that it says that's outdated?

aspiwack replied:
Mmm, I see. I misread.

aspiwack commented on src/Control/Applicative/Static.hs:
```diff
@@ -0,0 +1,33 @@
+--
+-- @
+-- staticCompose :: Closure (b -> c) -> Closure (a -> b) -> Closure (a -> c)
+-- staticCompose f g = static (.) ``cap`` f ``cap`` g
```
Thinking aloud: ~~what would it take/cost to define a `Category` instance for `Closure (_ -> _)`? Hence use `(.)` and `id` freely?~~ Oh you did the `newtype` below already!

mboes replied:
Yes, but didn't define this instance. Good idea!

mboes replied:
Actually, the instance can't be defined, at least not for `Category` in `base`, because of the `Typeable` constraints. :-/

aspiwack replied:
Oh. Indeed.

aspiwack commented on src/Control/Applicative/Static.hs:
```diff
@@ -0,0 +1,33 @@
+    :: (Typeable a, Typeable b)
+    => f (Closure (a -> b))
+    -> f a
+    -> f b
```
I'm pretty convinced that this is not the right signature and that we should have `f (a ->b) -> f a -> f b` instead.

Reasoning: what we expect of `staticApply` is to be able to chain it as in the laws above. But this doesn't actually work: `g :: f (Closure (a -> b -> c)`, `x :: f a`, `y :: f b`; `f <*> x :: f (b -> c)`, `f <*> x <*> y` is ill typed.

In particular the laws are ill typed.

mboes replied:
Yes.

aspiwack commented on src/Control/Comonad/Static.hs:
```diff
@@ -0,0 +1,31 @@
+--
+-- @
+-- 'staticExtend' f = 'staticMap' f . 'staticDuplicate'
+-- 'staticDuplicate' = 'staticExtend' (static 'id')
```
Why aren't there laws relating `staticExtend` and `staticMap`?

From semigroupoid:
- `extended f . extended g = extended (f . extended g)`
- `duplicated . duplicated = fmap duplicated . duplicated`
(they should be equivalent if `extend` and `duplicated` are related as in the existing laws)

mboes replied:
> Why aren't there laws relating `staticExtend` and `staticMap`?

Isn't that the first law?

aspiwack replied:
I don't know why I wrote `staticMap` since it's obviously irrelevant from the two laws I cited after.
 
That being said, the two laws that you propose relate `staticExtend` and `staticDuplicate`, but don't constrain either beyond that. In contrast to either of the laws I pilfered from `semigroupoid`(which are associativity laws).

_again, I have no idea why I mentionned `staticMap`..._

aspiwack commented on src/Control/Distributed/Closure.hs:
```diff
@@ -52,3 +58,12 @@ import Data.Constraint (Dict(..))
+-- higher-kinded types.
+newtype WrappedArrowClosure a b = WrapArrowClosure
+  { unwrapClosureArrow :: Closure (a -> b)
+  } deriving (Binary)
```
What about just `ArrowClosure`, or even `CArrow`?

You should define the `Category` instance for it too.

mboes replied:
For consistency with `base`, which defines e.g. `WrappedMonad` and `WrappedArrow` (the latter of which also being present in `profunctors`).

aspiwack commented on src/Control/Distributed/Closure/Class.hs:
```diff
@@ -0,0 +1,34 @@
+-- do is map /static pointers to/ arbitrary functions over it (or in general,
+-- closures). 'Closure' is not just an applicative functor, it's also a monad,
+-- as well as a comonad, if again we limit the function space to those functions
+-- that can be statically pointed to.
```
Duplicate of the comment in `Functor.Static`. Probably scratch one of these.

mboes replied:
Ah. This module was meant to be deleted entirely.

aspiwack commented on src/Control/Monad/Static.hs:
```diff
@@ -0,0 +1,43 @@
+-- respectively):
+--
+-- @
+-- (m >>= f) >>= g = m >>= static (.) ``cap`` (staticFlippedBind g) `cap` f
```
Interesting. `staticFlipBind` is the more general version of `staticBind` that I've mentioned before. You can't actually define it for an arbitrary `StaticBind m`, as the constraint is not known to be `Static`. 

(the same set of remarks holds for `staticJoin` below)

mboes replied:
> You can't actually define it for an arbitrary StaticBind m, as the constraint is not known to be Static.

No, but you can define it on a case by case basis. Is that what you mean?

aspiwack replied:
Right. You can for any closed type `m`. It's a bit fishy as it doesn't quite allow to express monad-transformer like things (their laws seem to only hold for static monad arguments).

It may be fine. But it may also suggest that the primitive method is something like
```haskell
   staticKleisliExtension :: (Typeable a, Typeable b) => Closure (a -> m b) -> Closure (m a -> m b)
```
From which `staticBind` has an obvious default implementation (`staticBind = flip (unclosure . staticKleisliExtension)`).

I don't know how much this matters, and how much I'm overthinking this :-p .

mboes replied:
We can recover one from the other for any concrete instance, so I wouldn't worry too much about it. ;)

aspiwack commented on src/Data/Profunctor/Static.hs:
```diff
@@ -0,0 +1,41 @@
+-- 'staticDimap' (static id) (static id) = static id
+-- 'staticLmap' (static id) = static id
+-- 'staticRmap' (static id) = static id
+-- 'staticDimap' f g = staticLmap f . staticRmap g
```
Kmett claims that, by parametricity, these laws imply the laws for composition. I haven't worked out how it works yet, so I don't know whether that holds for closures as well (though I assume it would).

mboes committed:
Remove C.D.C.Class module.

Holdover from previous commits.

mboes committed:
Remove Closure from definition of Applicative.

mboes commented:
PTAL

aspiwack reviewed (approved):
I believe it's all consistent now.

mboes committed:
Add extra Comonad laws.
